tiny_cnn/core/kernels/conv2d_tiny_op.h

```diff
@@ -44,36 +44,152 @@
 */
 #pragma once
 
+#include "tiny_cnn/core/kernels/conv2d.h"
+#include "tiny_cnn/core/kernels/conv2d_tiny_op_functor.h"
+
 #include "tiny_cnn/core/framework/op_kernel.h"
 
 namespace tiny_cnn {
 
-class Conv2dCustomForwardOp : public core::OpKernel {
+class Conv2dCustomForwardOp : public core::OpKernel, Conv2d {
  public:
     explicit Conv2dCustomForwardOp(core::OpKernelConstruction* context)
         : core::OpKernel(context) {}
 
     void compute(core::OpKernelContext* context) override {
-
-        nn_info("Running Conv2dCustomForwardOp");
-
-        // TODO(edgar/nyanp): we should be able to do something like this
-        //
-        // const tensor_t& in_data = context->input(0);
-        // const vec_t&          W = context->input(1);
-        // const vec_t&          b = context->input(2);
-        // tensor_t&      out_data = context->output(0);
-
+        // incomimg/outcoming data 
+        const tensor_t& in_data = context->input(0);
+        const vec_t&          W = context->input(1)[0];
+        const vec_t&       bias = context->input(2)[0];
+        tensor_t&      out_data = context->output(1);
+
+        // pad input data
+        tensor_t in_data_padded;
+        Conv2d::setParams(context->params());
+        Conv2d::copy_and_pad_input(in_data, in_data_padded);
+
+        // convolution algorithm
+        Conv2dCustomForwardFunctor()(in_data_padded, W, bias, out_data,
+                                     Conv2d::params(), context->parallelize());
```

Comment:
Well C++ is not an OOP language, it's a multi-paradigm language. We can choose the simplest way to handle the problem. If we'll have to hold the state, we can hold them in ```conv2d_tiny_op``` class. (And actually I don't agree that adding ```XXXFunctor``` for handling the state is good OOP style...)
Performance would be differ if we use them as predicate. In this case there is no difference between functor and normal function :)